group.go

```diff
@@ -21,202 +21,166 @@
-	"github.com/topfreegames/pitaya/util"
 )
 
 const (
 	groupStatusWorking = 0
 	groupStatusClosed  = 1
 )
 
-// SessionFilter represents a filter which is used to filter sessions when Multicast,
-// the session will receive the message when the filter returns true.
-type SessionFilter func(*session.Session) bool
-
-// Group represents a session group which is used to manage a number of
-// sessions, data sent to the group will be sent to all sessions in it.
+// Group represents an agglomeration of UIDs which is used to manage
+// users. Data sent to the group will be sent to all users in it.
 type Group struct {
-	mu       sync.RWMutex
-	status   int32                       // channel current status
-	name     string                      // channel name
-	sessions map[string]*session.Session // session id map to session instance
+	groupService groups.GroupService
+	status       int32  // channel current status
+	name         string // channel name
 }
 
 // NewGroup returns a new group instance
-func NewGroup(n string) *Group {
+func NewGroup(n string, gs groups.GroupService) *Group {
 	return &Group{
-		status:   groupStatusWorking,
-		name:     n,
-		sessions: make(map[string]*session.Session),
+		status:       groupStatusWorking,
+		name:         n,
+		groupService: gs,
 	}
 }
 
-// Member returns specified UID's session
-func (c *Group) Member(uid string) (*session.Session, error) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-
-	s, ok := c.sessions[uid]
-	if ok {
-		return s, nil
+// MemberGroups returns all groups that the member takes part
+func (c *Group) MemberGroups(ctx context.Context, uid string) ([]string, error) {
+	if c.isClosed() {
+		return nil, constants.ErrClosedGroup
+	}
+	if uid == "" {
+		return nil, constants.ErrNoUIDBind
 	}
-	return nil, constants.ErrMemberNotFound
+	return c.groupService.MemberGroups(ctx, uid)
 }
 
-// Members returns all member's UID in current group
-func (c *Group) Members() []string {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-
-	members := []string{}
-	for _, s := range c.sessions {
-		members = append(members, s.UID())
+// Member returns member payload
+func (c *Group) Member(ctx context.Context, uid string) (*groups.Payload, error) {
+	if c.isClosed() {
+		return nil, constants.ErrClosedGroup
 	}
-
-	return members
+	if uid == "" {
+		return nil, constants.ErrNoUIDBind
+	}
+	return c.groupService.Member(ctx, c.name, uid)
 }
 
-// Multicast  push  the message to the filtered clients
-func (c *Group) Multicast(route string, v interface{}, filter SessionFilter) error {
+// Members returns all member's UIDs in current group
+func (c *Group) Members(ctx context.Context) (map[string]*groups.Payload, error) {
 	if c.isClosed() {
-		return constants.ErrClosedGroup
+		return nil, constants.ErrClosedGroup
 	}
+	return c.groupService.Members(ctx, c.name)
+}
 
-	data, err := util.SerializeOrRaw(app.serializer, v)
-	if err != nil {
-		return err
+// Multicast  pushes  the message to the filtered clients
```

Comment:
@rodrigolck where is the filter function being applied here?

Reply:
Will remove it, since it is not applicable anymore